### ISSUE TITLE
Make behavior consistent between `plottime_interval` and `plotfile_interval`

### ISSUE THREAD

**Issue by chongchonghe** [quokka-astro/quokka#634]
**Describe the proposal**
Currently, when a positive `plottime_interval` is set, the first plot file is saved at t = plottime_interval. It would great if a plot file at t = 0 is also saved. This is good for debugging, and sometimes I want to make plots from the initial condition.

**Comment by BenWibking** [comment]
The first problem can be fixed by modifying this line: https://github.com/quokka-astro/quokka/blob/4c1a57659490ef59a5034602e5b9d6acaca070db/src/simulation.hpp#L683

I don't know how the second problem happens, since it calls `InitFromScratch` before writing out the first plotfile...

**Comment by chongchonghe** [comment]
Also, I just noticed that the last plt and chk files are also bugged. I set `stop_time = 4.8e-5, plottime_interval = 2.0e-6, checkpointtime_interval = 2.0e-6` in parameter file, but the last chk and plt folder is at `t=4.6000023122425903e-05`.

**Comment by BenWibking** [comment]
> Also, I just noticed that the last plt and chk files are also bugged. I set `stop_time = 4.8e-5, plottime_interval = 2.0e-6, checkpointtime_interval = 2.0e-6` in parameter file, but the last chk and plt folder is at `t=4.6000023122425903e-05`.

Does it reach this line?
https://github.com/quokka-astro/quokka/blob/7a597a57579780a8e0e25ee0220296de479ff4ea/src/simulation.hpp#L982

If you have `plotfile_interval = 0`, it won't execute this line and write out a final plotfile. I think this was the intended behavior of https://github.com/quokka-astro/quokka/pull/227. That is, I don't think it's a bug.

**Comment by BenWibking** [comment]
Although it would make more sense to write the checkpoint file at the end if either `checkpoint_interval` or `checkpointtime_interval` are set.

It may or may not make sense to output the plotfile at the end. Sometimes that is only desired at fixed time intervals, e.g. for making movies.

**Comment by BenWibking** [comment]
> Also, when plotfile_interval = 1 is set, the first snapshot saved is plt00000. This is good, but the content of this snapshot is not right. It does not have the initial condition correctly set.

@chongchonghe Can you create a separate issue for this bug?

**Comment by chongchonghe** [comment]
> Although it would make more sense to write the checkpoint file at the end if either `checkpoint_interval` or `checkpointtime_interval` are set.
> 
> It may or may not make sense to output the plotfile at the end. Sometimes that is only desired at fixed time intervals, e.g. for making movies.

I agree. It's not a bug. I guess what happened is related to the comparison between floating point numbers. A workaround, probably what you should do, is to set stop_time to N * plotfile_interval + a small number.

**Comment by chongchonghe** [comment]
> > Also, when plotfile_interval = 1 is set, the first snapshot saved is plt00000. This is good, but the content of this snapshot is not right. It does not have the initial condition correctly set.
> 
> @chongchonghe Can you create a separate issue for this bug?

Never mind. I double check that and found it was my mistake. I plotted the wrong field in my python script. Nothing wrong with plt00000.

**Comment by chongchonghe** [comment]
> > Although it would make more sense to write the checkpoint file at the end if either `checkpoint_interval` or `checkpointtime_interval` are set.
> > It may or may not make sense to output the plotfile at the end. Sometimes that is only desired at fixed time intervals, e.g. for making movies.
> 
> I agree. It's not a bug. I guess what happened is related to the comparison between floating point numbers. A workaround, probably what you should do, if you want a pltfile at stop_time, is to set stop_time to N * plotfile_interval + a small number.

**Comment by BenWibking** [comment]
I agree it would make sense to make the behavior consistent between `plotfile_interval` and `plottime_interval` (and likewise for  checkpoint and checkpointtime):

* To fix your original issue, unless plotfile output is completely disabled, it should always output a plotfile at t=0.
* Since `plottime_interval` was originally designed to avoid outputting at the end of a run (if it didn't coincide with a time interval), we could add an extra runtime parameter, e.g., `output_when_stopping`. It would make sense for this to be on by default. Then if users didn't want an extra output when stopping, they could turn it off.
